core/src/main/java/org/apache/iceberg/io/datafile/WriteBuilder.java

```diff
@@ -0,0 +1,340 @@
+import java.util.stream.IntStream;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.MetricsConfig;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SortOrder;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.deletes.EqualityDeleteWriter;
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.encryption.EncryptionKeyMetadata;
+import org.apache.iceberg.io.DataWriter;
+import org.apache.iceberg.io.DeleteSchemaUtil;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.ArrayUtil;
+
+/**
+ * Builder for generating one of the following:
+ *
+ * <ul>
+ *   <li>{@link FileAppender}
+ *   <li>{@link DataWriter}
+ *   <li>{@link EqualityDeleteWriter}
+ *   <li>{@link PositionDeleteWriter}
+ * </ul>
+ *
+ * The builder wraps the file format specific {@link AppenderBuilder}. To allow further engine and
+ * file format specific configuration changes for the given writer the {@link
+ * AppenderBuilder#build(AppenderBuilder.WriteMode)} method is called with the correct parameter to
+ * create the appender used internally by the file appender and the writers.
+ *
+ * @param <A> type of the appender
+ * @param <E> engine specific schema of the input records used for appender initialization
+ */
+public class WriteBuilder<A extends AppenderBuilder<A, E>, E> {
+  private final AppenderBuilder<A, E> appenderBuilder;
+  private final String location;
+  private final FileFormat format;
+  private PartitionSpec spec = null;
+  private StructLike partition = null;
+  private EncryptionKeyMetadata keyMetadata = null;
+  private SortOrder sortOrder = null;
+  private Schema rowSchema = null;
+  private int[] equalityFieldIds = null;
+
+  WriteBuilder(AppenderBuilder<A, E> appenderBuilder, String location, FileFormat format) {
+    this.appenderBuilder = appenderBuilder;
+    this.location = location;
+    this.format = format;
+  }
+
+  /**
+   * Sets the configurations coming from the table like {@link #schema(Schema)}, {@link #set(Map)}
+   * and {@link #metricsConfig(MetricsConfig)}.
+   */
+  public WriteBuilder<A, E> forTable(Table table) {
+    appenderBuilder.forTable(table);
+    return this;
+  }
+
+  /** Set the file schema. */
+  public WriteBuilder<A, E> schema(Schema newSchema) {
+    appenderBuilder.schema(newSchema);
+    return this;
+  }
+
+  /**
+   * Sets the engine specific schema for the input. Used by the {@link
+   * AppenderBuilder#build(AppenderBuilder.WriteMode)} to configure the engine specific converters.
+   */
+  public WriteBuilder<A, E> engineSchema(E engineSchema) {
+    appenderBuilder.engineSchema(engineSchema);
+    return this;
+  }
+
+  /** Set the file schema's root name. */
+  public WriteBuilder<A, E> named(String newName) {
```

Comment:
Is this useful anymore? I don't think there's any way to control it and we generally use `table`. Maybe we should just hard-code that instead of exposing an extra method.

Reply:
Removed